### PR TITLE
8342047: Create Template Assertion Predicates with Halt nodes only instead of uncommon traps

### PR DESCRIPTION
**Pull request by chhagedorn**
This patch replaces the creation of Template Assertion Predicates with uncommon traps with Halt nodes.

#### Initialized Assertion Predicates
These predicates ensure that control is properly folded when data is dying. They are **always true by design** and thus can never fail at runtime. We therefore put a halt node on the failing path.

#### Template Assertion Predicates
Only serve as templates to create Initialized Assertion Predicates from. They are never executed and are always removed after loop opts are over. Conceptionally, it does not matter whether the failing path uses an UCT or a halt node (or something else completely - I plan to have a separate "no-op" `TemplateAssertionPredicateNode` at some point which only falls through to the next node and does not have a failing path at all).

### Why Did we Use UCTs for Template Assertion Predicates?
When the concept of Assertion Predicates was first introduced, it only covered a few edge cases. It was quite straight forward to reuse existing Loop Predication code which creates new predicates from a Parse Predicate by copying it and merging the UCTs on the failing paths with a region node. This was done with `PhaseIdealLoop::create_new_if_for_predicate()`.

#### Missing UCTs for Predicates above Loops
Over time, we found more cases where we need to create Initialized Assertion Predicates from templates - including locations where we do not have Parse Predicates (and thus no safepoints). For example, when peeling one iteration off a loop with Parse Predicates, they will be kept at the peeled iteration and the remaining loop does not have any Parse Predicates anymore.

#### Missing UCTs to Create Template Assertion Predicates
Whenever we split a loop with Template Assertion Predicates, we also need to ensure that they are copied to all split loop versions. Since they rely on using UCTs, we also need to make sure that an UCT/safepoint is available to be used. However, this is not always the case (for example, after peeling an iteration off as described in the last section). As a result, we cannot easily establish new Template Assertion Predicates anywhere. One could think about faking an UCT or doing other special logic. But this seems rather fragile and could introduce quite some complexity - especially since we conceptionally don't even need to use UCTs at all for Template Assertion Predicates.

There is already some special logic for a main loop, where we create Template Assertion Predicates with a Halt node because there is no UCT available for the main loop. But this logic and implementation it is not easily reusable and we would need to keep supporting both formats with UCTs and halt nodes.

### Solution: Assertion Predicates with Halt Nodes only
As a simple solution to the problems described above, I propose to get rid of UCTs completely. This not only enables us to fix the remaining unresolved bugs where Assertion Predicates are missing but also simplifies the logic and the IR itself.

I've added some comments in the PR to better explain the refactoring steps.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342047](https://bugs.openjdk.org/browse/JDK-8342047): Create Template Assertion Predicates with Halt nodes only instead of uncommon traps (**Sub-task** - P4)

### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22040/head:pull/22040` \
`$ git checkout pull/22040`

Update a local copy of the PR: \
`$ git checkout pull/22040` \
`$ git pull https://git.openjdk.org/jdk.git pull/22040/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22040`

View PR using the GUI difftool: \
`$ git pr show -t 22040`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22040.diff">https://git.openjdk.org/jdk/pull/22040.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22040#issuecomment-2470854362)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 1]
```diff
@@ -287,58 +287,53 @@ IfProjNode* PhaseIdealLoop::clone_parse_predicate_to_unswitched_loop(ParsePredic
...
+        clone_assertion_predicate_for_unswitched_loops(template_assertion_predicate_success_proj,
+                                                       false_path_loop_parse_predicate);
 
     // Update control dependent data nodes.
-    for (DUIterator j = predicate->outs(); predicate->has_out(j); j++) {
-      Node* fast_node = predicate->out(j);
+    for (DUIterator j = template_assertion_predicate_success_proj->outs();
+         template_assertion_predicate_success_proj->has_out(j);
+         j++) {
+      Node* fast_node = template_assertion_predicate_success_proj->out(j);
```
Comment (eme64) [inline src/hotspot/share/opto/loopPredicate.cpp:314]:
```suggestion
      Node* true_path_node = template_assertion_predicate_success_proj->out(j);
```
Now that you have renamed it `fast_proj` -> `true_path_loop_proj`.
And why `true_path_loop_proj` and not just `true_path_proj`?

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 2]
```diff
@@ -287,58 +287,53 @@ IfProjNode* PhaseIdealLoop::clone_parse_predicate_to_unswitched_loop(ParsePredic
...
     // Have to delay updates to the slow loop so uses of predicate are not modified while we iterate on them.
     while (to_process.size() > 0) {
       Node* slow_node = to_process.pop();
-      _igvn.replace_input_of(slow_node, 0, slow_proj);
+      _igvn.replace_input_of(slow_node, 0, false_path_loop_proj);
     }
   }
 }
 
 // Put all Assertion Predicate projections on a list, starting at 'predicate' and going up in the tree. If 'get_opaque'
```
Comment (eme64) [inline src/hotspot/share/opto/loopPredicate.cpp:332]:
```suggestion
// Put all Template Assertion Predicate projections on a list, starting at 'predicate' and going up in the tree. If 'get_opaque'
```
Would that be more accurate?

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 3]
```diff
@@ -348,56 +343,56 @@ void PhaseIdealLoop::get_assertion_predicates(ParsePredicateSuccessProj* parse_p
...
+  set_idom(unswitched_loop_parse_predicate, template_success_proj, dom_depth(template_success_proj));
+  return template_success_proj;
 }
 
 // Clone the old Parse Predicates and Assertion Predicates before the unswitch If to the unswitched loops after the
 // unswitch If.
 void PhaseIdealLoop::clone_parse_and_assertion_predicates_to_unswitched_loop(IdealLoopTree* loop, Node_List& old_new,
-                                                                             IfProjNode*& iffast_pred, IfProjNode*& ifslow_pred) {
+                                                                             IfProjNode*& true_path_loop_entry,
+                                                                             IfProjNode*& false_path_loop_entry) {
```
Comment (eme64) [inline src/hotspot/share/opto/loopPredicate.cpp:362]: Oh boy... passing pointers by reference... I suppose that was already here like this. Looks adventurous :see_no_evil:

Comment (eme64) [reply src/hotspot/share/opto/loopPredicate.cpp:362]: Maybe there should be some sort of `UnswitchingResult`, that has the projections and `old_new` mapping? That could then be passed as a result. What do you think? Can be a separate RFE of course.

**Review by eme64 — CHANGES_REQUESTED** [review]
Reasonable :)
Looks like a nice simplicifaction that enables more improvements in the future!

**Review by eme64 — APPROVED** [review]
Thanks for the updates!

**Comment by chhagedorn** [comment]
Thanks Emanuel for your review!

File: src/hotspot/share/opto/loopPredicate.cpp (hunk 3 above)
Comment (TobiHartmann) [reply src/hotspot/share/opto/loopPredicate.cpp:362]: It's quite common to pass a pointer by reference in C2 code, unfortunately.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by chhagedorn** [comment]
Thanks Tobias for your review!

/integrate

**Pull request closed without merging**